src/MotorPlugin.cc

```diff
@@ -0,0 +1,548 @@
+
+	// Bounds checking
+	if (controlIndex >= 0 && controlIndex < static_cast<int>(this->velValues.size()))
+	{
+		this->velValues[controlIndex] = _msg.data();
+	}
+	else
+	{
+		gzwarn << "Invalid control index " << controlIndex << " for PWM message. Expected [0, " 
+			<< (this->velValues.size() - 1) << "]" << std::endl;
+	}
+}
+
+
+//////////////////////////////////////////////////
+MotorPlugin::~MotorPlugin() = default;
+
+//////////////////////////////////////////////////
+MotorPlugin::MotorPlugin() : impl(std::make_unique<MotorPlugin::Impl>())
+{
+}
+
+//////////////////////////////////////////////////
+void MotorPlugin::Configure(
+    const Entity &_entity,
+    const std::shared_ptr<const sdf::Element> &_sdf,
+    EntityComponentManager &_ecm,
+    EventManager &)
+{
+
+    // Make a clone so that we can call non-const methods
+    sdf::ElementPtr sdfClone = _sdf->Clone();
+
+    // retrieve world entity
+    this->impl->world = World(
+        _ecm.EntityByComponents(components::World()));
+    if (!this->impl->world.Valid(_ecm))
+    {
+        gzerr << "MotorPlugin - world not found. "
+                "Failed to initialize.\n";
+        return;
+    }
+    this->impl->worldName = this->impl->world.Name(_ecm).value();
+    
+    // capture model entity
+    this->impl->parentModel = Model(_entity);
+    if (!this->impl->parentModel.Valid(_ecm))
+    {
+        gzerr << "MotorPlugin should be attached to a model. "
+                "Failed to initialize.\n";
+        return;
+    }
+    this->impl->parentModelName = this->impl->parentModel.Name(_ecm);
+
+    // Load control channel params
+    this->LoadControlChannels(sdfClone, _ecm);
+
+    // Initialize msg values vector for safety
+    this->impl->velValues.resize(this->impl->controls.size(), 0.0);
+
+    // create components and subscriptions.
+    for (int i = 0; i < this->impl->controls.size(); ++i)
+    {
+        auto &control = this->impl->controls[i];
+        
+        // Create joint components
+        if (!_ecm.Component<gz::sim::components::JointVelocity>(control.joint))
+        {
+            _ecm.CreateComponent(control.joint, gz::sim::components::JointVelocity({0.0}));
+        }
+        if (!_ecm.Component<gz::sim::components::JointForceCmd>(control.joint))
+        {
+            _ecm.CreateComponent(control.joint, gz::sim::components::JointForceCmd({0.0}));
+        }
+
+        // Subscriber
+        std::string topic = this->impl->topics[i];
+        this->impl->node.Subscribe<gz::msgs::Double>(
+            topic,
+            [this, i](const gz::msgs::Double &_msg, const gz::transport::MessageInfo &_info) {
+                this->impl->OnVelMsg(i, _msg);
+            });
+
+        gzdbg << "MotorPlugin subscribing to PWM messages on [" << topic 
+            << "] for control index " << i << std::endl;
+    }
+    this->impl->validConfig = true;
+
+}
+
+/////////////////////////////////////////////////
+void MotorPlugin::LoadControlChannels(
```

Comment:
```suggestion
void MotorPlugin::Impl::LoadControlChannels(
```

And move declaration from `MotorPlugin.hh`.